### PR TITLE
make Publisher uses RetrySettings

### PR DESCRIPTION
**Pull request by pongad**
cc @davidtorres

### CODE REVIEW

File: google-cloud-pubsub/src/main/java/com/google/cloud/pubsub/spi/v1/Publisher.java [hunk 1]
```diff
@@ -150,14 +151,24 @@
...
             .build();
+    static final RetrySettings DEFAULT_RETRY_SETTINGS =
+        RetrySettings.newBuilder()
+            .setTotalTimeout(MIN_TOTAL_TIMEOUT)
+            .setInitialRetryDelay(Duration.millis(5))
+            .setRetryDelayMultiplier(2)
+            .setMaxRetryDelay(Duration.millis(Long.MAX_VALUE))
+            .setInitialRpcTimeout(DEFAULT_REQUEST_TIMEOUT)
+            .setRpcTimeoutMultiplier(2)
+            .setMaxRpcTimeout(DEFAULT_REQUEST_TIMEOUT)
```
Comment (pongad) [inline google-cloud-pubsub/src/main/java/com/google/cloud/pubsub/spi/v1/Publisher.java]: This is left the same with initial RPC timeout so preserve the old behavior that the value stays the same. I'm not too worried about it since it's easy to change later.

File: google-cloud-pubsub/src/main/java/com/google/cloud/pubsub/spi/v1/Publisher.java [hunk 2]
```diff
@@ -150,14 +151,24 @@
...
+    static final Duration MIN_RPC_TIMEOUT = new Duration(10); // 10 milliseconds
     static final BundlingSettings DEFAULT_BUNDLING_SETTINGS =
         BundlingSettings.newBuilder()
             .setDelayThreshold(DEFAULT_MAX_BUNDLE_DURATION)
             .setRequestByteThreshold(DEFAULT_MAX_BUNDLE_BYTES)
             .setElementCountThreshold(DEFAULT_MAX_BUNDLE_MESSAGES)
             .build();
+    static final RetrySettings DEFAULT_RETRY_SETTINGS =
+        RetrySettings.newBuilder()
+            .setTotalTimeout(MIN_TOTAL_TIMEOUT)
```
Comment (garrettjonesgoogle) [inline google-cloud-pubsub/src/main/java/com/google/cloud/pubsub/spi/v1/Publisher.java]: We should distinguish the concepts of minimum and default - make a new constant `DEFAULT_TOTAL_TIMEOUT` and use that here.

File: google-cloud-pubsub/src/main/java/com/google/cloud/pubsub/spi/v1/PublisherImpl.java [hunk 3]
```diff
@@ -150,17 +140,21 @@ public PublisherStats getStats() {
...
-    return maxBundleBytes;
+    return bundlingSettings.getRequestByteThreshold();
+  }
+
+  private boolean hasBundlingBytes() {
+    return getMaxBundleBytes() > 0;
   }
 
   @Override
   public long getMaxBundleMessages() {
```
Comment (garrettjonesgoogle) [inline google-cloud-pubsub/src/main/java/com/google/cloud/pubsub/spi/v1/PublisherImpl.java]: We should probably just have an accessor for bundling settings, because that is the type that is provided at build time.

File: google-cloud-pubsub/src/main/java/com/google/cloud/pubsub/spi/v1/Publisher.java [hunk 4]
```diff
@@ -150,14 +151,24 @@
     static final long DEFAULT_MAX_BUNDLE_BYTES = 1000L; // 1 kB
     static final Duration DEFAULT_MAX_BUNDLE_DURATION = new Duration(1); // 1ms
     static final Duration DEFAULT_REQUEST_TIMEOUT = new Duration(10 * 1000); // 10 seconds
```
Comment (garrettjonesgoogle) [inline google-cloud-pubsub/src/main/java/com/google/cloud/pubsub/spi/v1/Publisher.java]: This should be `DEFAULT_RPC_TIMEOUT` for consistency.

File: google-cloud-pubsub/src/main/java/com/google/cloud/pubsub/spi/v1/Publisher.java [hunk 5]
```diff
@@ -150,14 +151,24 @@
     static final long DEFAULT_MAX_BUNDLE_BYTES = 1000L; // 1 kB
     static final Duration DEFAULT_MAX_BUNDLE_DURATION = new Duration(1); // 1ms
```
Comment (garrettjonesgoogle) [inline google-cloud-pubsub/src/main/java/com/google/cloud/pubsub/spi/v1/Publisher.java]: Rename these other defaults to match the fields they are used to set (You can do this in a separate PR)

Comment (pongad) [inline google-cloud-pubsub/src/main/java/com/google/cloud/pubsub/spi/v1/Publisher.java]: Done. MIN_{TOTAL,RPC}_TIMEOUT is kept unchanged, since they are actually the minimum we allow. The default just also happens to be the min.

File: google-cloud-pubsub/src/main/java/com/google/cloud/pubsub/spi/v1/Publisher.java (hunk 4 above)
Comment (pongad) [inline google-cloud-pubsub/src/main/java/com/google/cloud/pubsub/spi/v1/Publisher.java]: Done

File: google-cloud-pubsub/src/main/java/com/google/cloud/pubsub/spi/v1/Publisher.java (hunk 2 above)
Comment (pongad) [inline google-cloud-pubsub/src/main/java/com/google/cloud/pubsub/spi/v1/Publisher.java]: Done

File: google-cloud-pubsub/src/main/java/com/google/cloud/pubsub/spi/v1/PublisherImpl.java (hunk 3 above)
Comment (pongad) [inline google-cloud-pubsub/src/main/java/com/google/cloud/pubsub/spi/v1/PublisherImpl.java]: I'd like to defer this for when we make Publisher a concrete class. Then we'll have fewer places to change.

**Comment by pongad** [comment]
@garrettjonesgoogle PTAL

File: google-cloud-pubsub/src/main/java/com/google/cloud/pubsub/spi/v1/PublisherImpl.java (hunk 3 above)
Comment (garrettjonesgoogle) [inline google-cloud-pubsub/src/main/java/com/google/cloud/pubsub/spi/v1/PublisherImpl.java]: sounds good.

**Comment by garrettjonesgoogle** [comment]
LGTM - but put the failure here on your list to investigate: https://travis-ci.org/GoogleCloudPlatform/google-cloud-java/jobs/192903764 (it is a pubsub failure; it seems intermittent since 3 builds passed, this one failed)

**Comment by pongad** [comment]
As discussed, we can merge to pubsub-hp, but we'll definitely hold off merging to master until we fix the flakes.
